plugins/query-insights/src/main/java/org/opensearch/plugin/insights/rules/model/SearchQueryRecord.java

```diff
@@ -0,0 +1,155 @@
+     *
+     * @return the observation time in milliseconds
+     */
+    public long getTimestamp() {
+        return timestamp;
+    }
+
+    /**
+     * Returns the measurement associated with the specified name.
+     *
+     * @param name the name of the measurement
+     * @return the measurement object, or null if not found
+     */
+    public Measurement<Number> getMeasurement(MetricType name) {
+        return measurements.get(name);
+    }
+
+    /**
+     * Returns an unmodifiable map of all the measurements associated with the metric.
+     *
+     * @return an unmodifiable map of measurement names to measurement objects
+     */
+    public Map<MetricType, Measurement<Number>> getMeasurements() {
+        return measurements;
+    }
+
+    /**
+     * Returns an unmodifiable map of the attributes associated with the metric.
+     *
+     * @return an unmodifiable map of attribute keys to attribute values
+     */
+    public Map<Attribute, Object> getAttributes() {
+        return attributes;
+    }
+
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, ToXContent.Params params) throws IOException {
+        builder.startObject();
+        builder.field("timestamp", timestamp);
+        for (Map.Entry<Attribute, Object> entry : attributes.entrySet()) {
+            builder.field(entry.getKey().toString(), entry.getValue());
+        }
+        for (Map.Entry<MetricType, Measurement<Number>> entry : measurements.entrySet()) {
+            builder.field(entry.getKey().toString(), entry.getValue().getValue());
+        }
+        return builder.endObject();
+    }
+
+    /**
+     * Write a SearchQueryRecord to a StreamOutput
+     * @param out the StreamOutput to write
+     * @throws IOException IOException
+     */
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeLong(timestamp);
+        out.writeMap(
+            measurements,
+            (stream, metricType) -> MetricType.writeTo(out, metricType),
+            (stream, measurement) -> Measurement.writeTo(out, measurement)
+        );
+        out.writeMap(attributes, (stream, attribute) -> Attribute.writeTo(out, attribute), StreamOutput::writeGenericValue);
+    }
+
+    /**
+     * Compare two SearchQueryRecord, based on the given MetricType
+     *
+     * @param a the first SearchQueryRecord to compare
+     * @param b the second SearchQueryRecord to compare
+     * @param metricType the MetricType to compare on
+     * @return 0 if the first SearchQueryRecord is numerically equal to the second SearchQueryRecord;
+     *        -1 if the first SearchQueryRecord is numerically less than the second SearchQueryRecord;
+     *         1 if the first SearchQueryRecord is numerically greater than the second SearchQueryRecord.
+     */
+    public static int compare(SearchQueryRecord a, SearchQueryRecord b, MetricType metricType) {
+        if (metricType == MetricType.LATENCY) {
```

Comment:
Ideally, `Measurement` value should be a `Comparable` interface and you can avoid if else checks here.